### ISSUE TITLE
Batch Jacobian like tf.GradientTape

### ISSUE THREAD

**Issue by Gsunshine** [pytorch/pytorch#23475]
## 🚀 Feature

We hope to get a **parallel** implementation of the **batch** Jacobian calculation like tensorflow, e.g. 

```
from tensorflow.python.ops.parallel_for.gradients import jacobian
jac = jacobian(y, x)
```

```
with tf.GradientTape() as g:
  x = tf.placeholder(tf.float32,  shape=(None, 3, 32, 32))
  g.watch(x)
  y = f(x)

batch_jacobian = g.batch_jacobian(y, x)
```

**Comment by colesbury** [comment]
@Gsunshine what does parallel mean in this context?

No one is currently working on this, but I think we'd happily accept a PR.

**Comment by Gsunshine** [comment]
A toy implementation embeded in Python. 

```
def jacobian_in_batch(y, x):
    '''
    Compute the Jacobian matrix in batch form.
    Return (B, D_y, D_x)
    '''

    batch = y.shape[0]
    single_y_size = np.prod(y.shape[1:])
    y = y.view(batch, -1)
    vector = torch.ones(batch).to(y)

    # Compute Jacobian row by row.
    # dy_i / dx -> dy / dx
    # (B, D) -> (B, 1, D) -> (B, D, D)
    jac = [torch.autograd.grad(y[:, i], x, 
                               grad_outputs=vector, 
                               retain_graph=True,
                               create_graph=True)[0].view(batch, -1)
                for i in range(single_y_size)]
    jac = torch.stack(jac, dim=1)
    
    return jac
```

However, it relies on a for loop in Python with `O(D_y)` complexity. This is based on the reverse mode AD with API torch.autograd.grad. For a minimum modification to Pytorch core, it is reasonable to compute the Jacobian matrix via a parallel for loop, like `pfor` in tensorflow. Otherwise, we need to change the autograd mechanism in pytorch, which is expensive and unreasonable.

The question is how difficult it is to add `pfor` for pytorch. It seems not easy according to the implementation in tensorflow. But I think it is a desirable faeture. If the complexity is well-estimated, I will implement it and send a PR.

Any suggestions? Thx!

**Comment by chanchann** [comment]
Great suggestion! Pytorch should have an efficient implementation of jacobian.

**Comment by soumith** [comment]
A naive "parallel" can simply be a python for-loop.

A tf `pfor` basically masks, does the op in parallel and unmasks, and it only supports a limited set of operators "parallely" that they hand-registered.

The closest thing we have in-flight with the same rough concept is [NestedTensor](https://github.com/pytorch/pytorch/issues/22169) which is a bit more general than `pfor`, and is similar to `tf`'s `RaggedTensor`

The similar but other instances in literature to do this are for example [Dynet's dynamic batching](https://github.com/pytorch/pytorch/issues/1642).

An alternate approach is to thread all of these parallely (using python threads or multiprocessing) and hope for good inter-op / intra-op parallelism setup in the framework, like discussed [here](https://github.com/pytorch/pytorch/issues/19001).

This is not a simple issue to work on.

**Comment by apaszke** [comment]
FWIW if we ramped up our symbolic autodiff in the JIT we could implement a JAX-like wrapper around PyTorch, and then this would be expressible. Otherwise it would be quite hard to transform all our derivatives to work in a batched setting (again, unless we move them to TorchScript, where this batching transform can be applied automatically).

**Comment by Gsunshine** [comment]
Note my toy implementation above. A for-loop is applied to y's dimension, not the batch dimension. And I test it in a toy case. It seems its complexity increases not linearly.

```
In [82]: x = torch.randn([128, 3, 32, 32]).requires_grad_()

In [83]: conv = nn.Conv2d(3, 1, 32, bias=False)

In [84]: y = conv(x).view(128, -1)

In [85]: y.shape
Out[85]: torch.Size([128, 1])

In [86]: timeit batch_jacobian(y, x)
4.31 ms Â±1.36 ms per loop (mean Â±std. dev. of 7 runs, 100 loops each)

In [87]: conv = nn.Conv2d(3, 128, 32, bias=False)

In [88]: y = conv(x).view(128, -1)

In [89]: y.shape
Out[89]: torch.Size([128, 128])

In [90]: timeit batch_jacobian(y, x)
3.29 s Â±227 ms per loop (mean Â±std. dev. of 7 runs, 1 loop each)
```

@soumith Besides, I test some cases in `tf`'s `pfor`. A basic observation is that they do not support `tf.cond` and `tf.while_loop`. Thus BN is not available. I am trying to follow #22169 and #22783 . However, it seems there is not vectorized support in the current prototype.

@apaszke Is it possible now to accelerate it via current `torch.jit.script`?

I agree it is actually not a simple issue to work on.

**Comment by apaszke** [comment]
The non-linear behavior is likely not because of the differentiation logic, but e.g. Conv2d picking different algorithms internally when you change the dimensionality. If you want to see the true complexity without any hidden variables like that use a simpler function (e.g. a polynomial).

I actually have an idea for how to do batched autograd, and it might not even be too complicated. I'll see what we can do.

**Comment by Gsunshine** [comment]
@apaszke It is a greatly desirable feature. We are very appreciate if you can help.

Is it possible that we can hack a faster implementation than a naive for-loop, if is might not be too complicated?

**Comment by zakizhou** [comment]
> FWIW if we ramped up our symbolic autodiff in the JIT we could implement a JAX-like wrapper around PyTorch, and then this would be expressible. Otherwise it would be quite hard to transform all our derivatives to work in a batched setting (again, unless we move them to TorchScript, where this batching transform can be applied automatically).

Is there a timeline for this feature? It's great to have explicit hessian and jacobian in pytorch like jax.

**Comment by apaszke** [comment]
Explicit jacobian and hessian are still relatively easy to do as shown [in this gist](https://gist.github.com/apaszke/226abdf867c4e9d6698bd198f3b45fb7). We should upstream those functions so that people don't have to write them out by hand though. In the future we'll also be able to compute batched gradients, thanks to the nested tensor patches that @cpuhrsch is working on.

**Comment by n-gao** [comment]
This feature would be great!
Is there any faster way than a for loop atm? I have to compute the derivatives of every network output (which is one per sample in a batch) wrt every parameter in the network. So, if my batch_size is N and I have M parameters I need a NxM matrix. The batch size has to be quite large and this has to be evalauted at every training step. The for loop is just not fast enough for this purpose.

**Comment by tsauri** [comment]
I need to get jacobian of weights wrt outputs. Can weigth.grad be expanded?
Currently I backware n output times and zero_grad manually

**Comment by radenmuaz** [comment]
> FWIW if we ramped up our symbolic autodiff in the JIT we could implement a JAX-like wrapper around PyTorch, and then this would be expressible. Otherwise it would be quite hard to transform all our derivatives to work in a batched setting (again, unless we move them to TorchScript, where this batching transform can be applied automatically).

Any updates on this issue?

**Comment by cpuhrsch** [comment]
@radenmuaz - [we're working on it](https://github.com/pytorch/nestedtensor).

**Comment by zou3519** [comment]
> Explicit jacobian and hessian are still relatively easy to do as shown [in this gist](https://gist.github.com/apaszke/226abdf867c4e9d6698bd198f3b45fb7). We should upstream those functions so that people don't have to write them out by hand though. In the future we'll also be able to compute batched gradients, thanks to the nested tensor patches that @cpuhrsch is working on.

cc @albanD how do you feel about upstreaming these for now so that users don't have to write them out by hand? I see requests for those from time-to-time ([this](https://discuss.pytorch.org/t/how-to-penalize-norm-of-end-to-end-jacobian/62771/16) is the most recent one) and each time I write out the code so I think it would be nicer to have those as helper functions even though they may not be the most efficient things in the world.

**Comment by albanD** [comment]
Yes, I want to do it: https://github.com/pytorch/pytorch/issues/30632

**Comment by patrick-kidger** [comment]
Relatedly, any hope for a batch-vjp? Better autodiff is a big reason I've been looking at moving to JAX.

**Comment by albanD** [comment]
Yes computing Jacobians in batch is done with batch-vjp. So yes this is in progress as well.

**Comment by zou3519** [comment]
@patrick-kidger, what exactly do you mean by batch-vjp? Are you looking for something that:
- computes vjp but can take in a "batch" of v's (so we effectively get a matrix-jacobian product)
- or do you have a batch of examples and want to compute the vjp for each example?

**Comment by patrick-kidger** [comment]
@zou3519 In my case, the latter, i.e. to align the batch dimensions of both vector and Jacobian.

**Comment by zou3519** [comment]
The latter is more https://github.com/pytorch/pytorch/issues/7786 I think. I'm curious about your use case, though. Do you have an example of the quantity that you'd like to compute or a reference/paper/code somewhere that we could take a look at?

**Comment by patrick-kidger** [comment]
Hmm, yeah that does look similar. Not sure it provides any solutions, though.

This is something that has come up as part of our development of [`torchsde`](https://github.com/google-research/torchsde). Certain derivatives are needed during stepping for SDE solvers. Moreover there shouldn't be any interaction between batch elements as the current autograd mandates.

For the purposes of this discussion, what we need to compute is
```
sum_{j, l} dg_{i,l}/dx_j B_{j, l}   (the index i is not summed over)
```
for some matrix-valued function `g`, and a matrix `B`. That is to say, a batched Jacobian-matrix contraction. Moreover there are actually two levels of batching: over the index `i` that is shown, and implicitly throughout the whole computation (over both `g` and `B`), which is over the actual batch elements in the usual sense.

In an ideal world, given a more powerful autograd, we'd use autograd to do a batch vjp or jvp to contract along either `j` or `l` (I think whichever is larger should be the more efficient option.) and then just manually sum up along the other `l` or `j`. (With several levels of batching, in one form or another - over the other `l` or `j`; over `i`; over the implicit batch dimension throughout.)

You can see the code, including both of our current implementations, here -

https://github.com/google-research/torchsde/blob/d1db41a56d01d01be9507f339382b0ea1e8d0833/torchsde/_core/base_sde.py#L111

You can see that the two ways we've got of doing it at the moment involves a smorgasbord of tricks, the composition of which is actually pretty far from what's theoretically achievable. (And could perhaps also still be improved? Not that we could see at the time, at least - one thing or another always seemed to go wrong.)

I'm happy to go into this in more detail if you're curious, but this is the main idea.

(CC @lxuechen for interest)

**Comment by soulitzer** [comment]
Closing this because we have many ways of doing this now:
- `torch.autograd.functional.jacobian(vectorized=True` which uses the vmap feature currently in core.
- `torch.autograd.grad(is_grads_batched=True` for more general batched gradient computation which also uses the vmap in core.
- For those interested in trying out [functorch](https://github.com/facebookresearch/functorch ) we also have `jacrev` and `vmap(vjp`

**Comment by dimikout3** [comment]
> Closing this because we have many ways of doing this now:
> 
> * `torch.autograd.functional.jacobian(vectorized=True` which uses the vmap feature currently in core.
> * `torch.autograd.grad(is_grads_batched=True` for more general batched gradient computation which also uses the vmap in core.
> * For those interested in trying out [functorch](https://github.com/facebookresearch/functorch) we also have `jacrev` and `vmap(vjp`

I am currently having a hard time to fully understand and use the `torch.autograd.functional.jacobian(vectorized=True`

I want to implement a batched jacobian like:

```
def jacobianBatch(f, x):
    '''
    Compute the jacobian (derivaties of outputs w.r.t inputs)
    Input:
            f: pytorch model
            x: batch of training data
    Output:
            jacobian: J of the batch of training data
    '''
    jacobian = []
    for i in range(x.shape[0]):
        jac = torch.autograd.functional.jacobian(f, x[i])
        jacobian.append(jac)

    return torch.stack(jacobian, 0)
```
Can anyone please provide a more efficient approach (for-loop requires a lot of time ...).
Thanks in advance!

**Comment by zou3519** [comment]
@dimikout3 check out functorch: https://github.com/pytorch/functorch. The `vmap` feature is designed to get rid of for-loops like this :)

We could rewrite your `jacobianBatch` function as so:
```
from functorch import jacrev, vmap

def jacobianBatch(f, x):
  return vmap(jacrev(f))(x)
```

**Comment by dimikout3** [comment]
@zou3519  Great, it works! Thanks a lot! Keep up the awesome work! The function runs incredibly faster now (as expected).

**Comment by zou3519** [comment]
> @zou3519 Great, it works! Thanks a lot! Keep up the awesome work! The function runs incredibly faster now (as expected).

@dimikout3 Out of curiosity, and if you can share, what are you using batch jacobian for? (and how were the speedups?)

**Comment by dimikout3** [comment]
@zou3519 We are currently working on a Deep Kalman Filter approach, the speedups are really impressive (for our tests we are having an improvement of something like x20, hopefully I will be able to share more soon).

**Comment by chaopengshen** [comment]
We need to compute jacobian in each time step, making it the slowest component if n_timestep is large.

I suspect if we can capture the graph and essentially make the jacobian itself static, we can make this part a lot faster. Any tools that can capture a static graph with jacobian and reuse it in pytorch?

**Comment by zou3519** [comment]
@chaopengshen maybe you can use torch.fx.experimental.proxy_tensor.make_fx to do this
